MetaMask Plugin/Tutorial.md

```diff
@@ -0,0 +1,244 @@
+
+![](Images/m5.png)
+
+
+`yarn start` will auto rebuild MetaMask on any file change. You can then [add your custom build to Chrome](https://metamask.zendesk.com/hc/en-us/articles/360016336611-Revert-Back-to-Earlier-Version-or-Add-Custom-Build-to-Chrome).
+
+You now have the forked the metamask on your machine.
+
+### Running Snap Dapps
+
+For building and running the snaps. Metamask provides us with the utility [snaps-cli](https://github.com/MetaMask/snaps-cli). 
+
+Install snaps-cli:
+
+```shell
+git clone https://github.com/MetaMask/snaps-cli
+cd snaps-cli
+npm i -g snaps-cli
+```
+To check the tools provided by snap-cli, run `mm-snap --help`.
+
+## Initializing
+
+Metamask has provided with the some examples in this [folder](https://github.com/MetaMask/snaps-cli/tree/master/examples).
+
+Let's start with the simplest one, `hello-snaps`
+
+```shell
+cd examples/hello-snaps
+mm-snap build
+mm-snap serve
+```
+
+`mm-snap build`: , build snap from sources into bundle.js
+`mm-snap serve`: Locally serve Snap file(s) for testing  
+
+
+This should give you a message `Server listening on: http://localhost:8081`. That port, and the build target are configured in the `snap.config.json` file, or command line arguments. You can now visit that address in your browser, and if you have installed your Snap branch of MetaMask correctly, you should be able to:
+
+* Click the "Connect" button on the site.
+* Approve the site's permissions request (which includes the Snap installation!)
+* Approve the Snap's permissions request (which in this case is permission to show alerts to you,   to send its message)
+* Click the "Send Hello" button to receive a greeting from the Snap.
+
+
+
+### Abstract level code description.
+
+File Structure :- 
+Dist
+Bundle.js - bundled js file of snap
+Index.js - snap code
+Package.json - Permissions are put here
+Package.-lock.json
+Snap.config.json - Snap configuration 
+ 
+The permissions the Snap initially requests are configured in its `package.json`, under the `web3Wallet key`.
+
+`index.html` file, interacts with the Snap using two basic API calls. It contains the DApp code. 
+In the `index.js` file, you can add API methods to connected websites from within a Snap. It contains the Snap Code.
+
+#### Snap Code `index.js`
+```
+wallet.registerRpcMessageHandler(async (originString, requestObject) => {
+  switch (requestObject.method) {
+    case 'hello':
+      return wallet.send({
+        method: 'alert',
+        params: [`Hello, ${originString}!`]
+      })
+    default:
+      throw new Error('Method not found.')
+  }
+})
+```
+
+The code registers a RPC Handler i.e., creates a developer defined API by the name of `hello` which can be called via the frontend given below. 
+
+- `requestObject` contains the method to be executed
+- `Alert` is an inbuilt method which allows us to create an alert on the webpage. 
+- An alert is created when the hello method is called using the metamask api.
+
+#### Dapp Code :- `index.html`
```

Comment:
@thesachinmittal This is not the whole file contents, either post everything, or mention that it's not the whole file

Reply:
@ChrisChinchilla Should I mention "Snippet from index.html"?